### DISCUSSION TITLE
My own research

### DISCUSSION THREAD

**Discussion by locatellidaniel** [THGSCST/Vices-and-Sins-MOD#1] [Discovery and Findings]
Hi, just came about with this project. Nice approach!

I've dedicated some time to research this game too. And more or less achieved similar results as yours.

So here are some questions:
* In StreetView. Did you have any success with extracting the "bigger" 192x128 tiles? I mean the tiles are 32x32, but you need 15 to make a big tile. Which is what is used in game. 
Probably you also found that the order is as follows:
```
       2
     1   3
   5   7    9
 4   6    8
  10   12   14
    11    13
       15
``` 
I can share a pic later maybe to explain this better. The thing is I couldn't find the relation between the header chunk for each tile to the position on this "bigger" tile. So my main suspicion is that somewhere else those tiles are defined, yet couldn't find where.

* Did you have any success with the Sprites.dat files?

* I see you also found the DebugFlag in the executable. I made a lot of effort trying to locate the parts of the code where the cases are handled, and how to solve that infamous bug of none of your hoods getting the proper judicial trial. But ran into a wall, it's very hard to rebuild the C structures used on those part of the code from the assembly code. Did you go any further with that?

I'm very happy to fond out that someone else dedicated time to this great game 👍

**Comment by THGSCST** [comment]
Greetings,

It's good to know that there are others also trying to analyse this game!

About StreetView sprites, I think you meant `locations.tyl` and `locations.dat` files. Initially, my main focus was just extracting and replacing the sprites, so I created the **SPRTYL2PNG** tool. If you open `locations.tyl` in the tool you will see all StreetView sprites (32x32 tiles), but they will be scrambled. This order you presented seems to be correct, but there are other layouts besides this one. I believe what you are looking for is in the `locations.dat` file. I haven't had time to fully look at it yet, but if you make any progress on this I can update the tool so an entire building can be displayed in the correct order. The tool also shows the tile index (non zero index), this might help a bit.
- [ ] TODO: Improve the SPRTYL2PNG tool.

The jail bug is complex, because to test it you would have to pay a bribe to the judge and the District Attorney and perhaps kill or intimidate some witnesses, this means loads of code and logic with variables to track. My biggest difficulty is debugging (using IDA or x64dbg). When I get to any breakpoint my video freezes, I need to find a way to run this game in windowed mode. So I couldn't go very far.

Another bug I was looking at is the hidden "Return to base" button and the removal of all cheats. I even managed to remove the DART and ILOVEHANSON cheats, but the removal of the button is not finished yet.

All the best!

**Comment by locatellidaniel** [comment]
Yes! I meant those two files: locations.dat eludes me. Last time I did some work was a few years ago so I don't remember very well. Maybe I can try again one of these days, with a fresh mind.

I didn't look at the Return to base button, what is the bug about? 

**Window debug**
I use dxwnd with this profile to run in a window. The menu doesn't show text, but the game works well. 
Debugging with IDA is possible, but with some caveats. You should hook the debugger after the game gets to the main menu (otherwise you get an "input sth" error), and sometimes the mouse lags when it stops after a breakpoint. Other than that I find it "good enough".

(note: I don't think the registry changes are necesary. I think I was trying sth and just left it like that. Replace XXX, YYY with the game path)
```
[target]
title0=Gangsters 1 (1.4)
path0=C:\XXXXXXXXX\gangsters.exe
startfolder0=C:\XXXXXXXXX\
launchpath0=
icon0=0000010001002020100000000000E802000016000000280000002000000040000000010004000000000080020000000000000000000000000000000000000000000000008000008000000080800080000000800080008080000080808000C0C0C0000000FF0000FF000000FFFF00FF000000FF00FF00FFFF0000FFFFFF00000000000087F8F70007F0000000000000000000007FFF700000F000000000000000000000FFFF00077700000000000000000000078F8788788FF000000000000000000008870788878FF80000000000000000000787878888770700000000000000000008F0F87888877077FFF00000000000000F878F88888888000007707700000000078088F888887000000FF0FF00000000078778888FF87880000FF0FFF0F00000070777877777778000000000F0F0000000870777777777700007707700000000088007708F007000000FF0FF000000007870077700000000000FF0FFF0F00000780000000000000000000000F0F00000070000000000000000000077000000000000000000000000000000FF000000000000000000FFFFF8000000FFF0F07000000000088FFFFFF880000000F0F07000000008888000000F80007077000700000008880000000000800FF0FF000700000778000000000000000FF0FFF0F70000077000000000000000000000F0F700007700000000000000000000000000000070000000000000000000000000000000000000000000000000077077F0F000000000000000000000000FF0FFF0F000000000000000000000000FF0FF0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001C0000000C004000040060000400700004006000000060000000600000006000000060000000400000000000000000000000000000000000000000000400000004008000040080000400C0000400E0000400F8000400F8001F04F8003F7FFC00FF7
module0=
opengllib0=
notes0=
registry0=[HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Hothouse\Gangsters]\n"PATH"="C:\\XXXXXXXXX\\YYYYYYYYY"\n"FirstTime"=dword:00000001\n"DRIVE"=dword:00000000\n"MIXER_SFX_VOLUME"=dword:000000ff\n"MIXER_CD_VOLUME"=dword:000000ff\n"MIXER_SFX_MUTE"=dword:00000000\n"MIXER_CD_MUTE"=dword:00000000\n"RandomPlay"=dword:00000000\n"ContPlay"=dword:00000001\n"RepeatPlay"=dword:00000000\n"GameAmbient"=dword:00000001\n"InterfaceAmbient"=dword:00000001\n"CurrentTrack"=dword:00000002\n"RepeatTrack"=dword:00000001\n"Checksum"=dword:89654ff8\n"DEFAULT_GENDER"=dword:00000000\n"DEFAULT_ETHNICGROUP"=dword:00000000\n"DEFAULT_HEAD"=dword:00000000\n"DEFAULT_HAIR"=dword:00000000\n"DEFAULT_EYES"=dword:00000000\n"DEFAULT_NOSE"=dword:00000000\n"DEFAULT_MOUTH"=dword:00000000\n"DEFAULT_COLOUR"=dword:00000000\n"DEFAULT_FIRSTNAME"="Louie"\n"DEFAULT_SECONDNAME"="Hothouse"\n"DEFAULT_NICKNAME"="Drake"\n"DEFAULT_GAMETYPE"=dword:00000000\n"DEFAULT_SGPOWER"=dword:00000014\n"DEFAULT_SGWEALTH"=dword:000186a0\n"DEFAULT_SGTERRITORY"=dword:00000064\n"DEFAULT_SCREENRES"=dword:00000400\n"DEFAULT_ECONOMY"=dword:00000001\n"DEFAULT_OPPONENTS"=dword:00000001\n"DEFAULT_MONEY"=dword:00001770\n"DEFAULT_HOODS"=dword:00000006\n"DEFAULT_TERRITORY"=dword:00000001\n"DEFAULT_MULTIPLAYER_GAMETYPE"=dword:00000000\n"DEFAULT_MULTIPLAYER_SGPOWER"=dword:00000014\n"DEFAULT_MULTIPLAYER_SGWEALTH"=dword:000186a0\n"DEFAULT_MULTIPLAYER_SGTERRITORY"=dword:00000064\n"DEFAULT_MULTIPLAYER_SCREENRES"=dword:00000400\n"DEFAULT_MULTIPLAYER_ECONOMY"=dword:00000001\n"DEFAULT_MULTIPLAYER_OPPONENTS"=dword:00000003\n"DEFAULT_MULTIPLAYER_MONEY"=dword:00001770\n"DEFAULT_MULTIPLAYER_HOODS"=dword:00000006\n"DEFAULT_MULTIPLAYER_TERRITORY"=dword:00000001\n"SP_SOUND"=dword:00000001\n"SP_CD"=dword:00000001\n"SP_VSYNC"=dword:00000001\n"SP_MMX"=dword:00000001\n"SP_MOUSE"=dword:00000000\n\n
ver0=1
monitorid0=-1
filterid0=0
renderer0=3
coord0=0
flag0=673185824
flagg0=1207959552
flagh0=1044
flagi0=1212153860
flagj0=4224
flagk0=589888
flagl0=4
flagm0=0
flagn0=17825796
flago0=0
tflag0=0
dflag0=0
posx0=50
posy0=50
sizx0=1024
sizy0=768
maxfps0=0
initts0=0
winver0=1
maxres0=-1
swapeffect0=0
maxddinterface0=7
slowratio0=2
scanline0=0
initresw0=800
initresh0=600
fakec0=
faked0=
cdvol0=100
[stamp]
version=v2.04.79.a1
time=Tue Jan 11 11:52:13 2022


```

**About the Jail bug**
The code that handles the cases is easy to locate because is full of logs. My guess is that they been trying to fix this up to the last minute. Also the DRM is around that area (pirate version send all your hoods to jail when you zoom the map).
I remember playing around and was able to make all the cases (even extortion) get through proper process, unfortunately after a few weeks the game hanged... guess it was too much to keep track :)
I bought old versions of the game (1.0) that didn't have the bug, to compare those routines. Yet this assembler and decompiling is not my main strength so i couldnt locate the right solution... yet¿?

As far as I remember the loop is like this:
* During Week phase, when a crime is comited a CrimeEvent is created and assigned an ID.
* At the end of the week, and before you get to the newspaper, a routine is called that goes through each open CrimeEvent and checks what to do next. Dropped cases are deleted here.

1. One week after a CrimeEvent is created. the "Investigation" phase starts.
2. Investigation can last several weeks, at the end it can be: solved "Crime solved!" or dropped. 
3. Drop reasons include: 
  * investigator was bribed (employ police) or maybe the Police chief has some influence.
  * perpetrator is dead
  * "Run out of time". 
4. Then the perpetrator is flagged for arrest.
5. Once in prison jury, judge, da are assigned. The event "goes to court"
6. In court several things can make the case to drop. No judges, DA available, bribed judged, etc.
7. Court lasts for a few weeks, then conviction or release happens.

I found two potential source of the bug.
* the "run out of time" during an investigation happens way too often. I'm not sure if this is related to some constants FEAR or sth like that. Tried with more policeman but didnt make a difference.
* the CrimeEvent never "goes to court" after your hood has been arrested so it stays there forever. No further events are logged, is like they're forgotten. Yet this could be related to the previous Investigation phase, I'm still not sure. Like some flag missing form the previous step.

**Reply by THGSCST** [reply]
That is great! You've already gone far beyond me in this matter. As soon as I have some free time I will look at it again.

About the "Return to base" button: for testing purposes the developers left (or forgot to disable) a hidden button that if pressed any neutral building can be taken over. In the popup window, you will see only the tip text, it is located here:
![image](https://user-images.githubusercontent.com/18038961/149249863-f9393d71-1c13-4ca9-a50c-958e37b07247.png)
This only works in the GangOrganizer screen. Even the Church building can be taken by the player.
